File: solutions/python/pig-latin/2/pig_latin.py

def translate(text):
    result = []
    for word in text.split():
        if word[0] in "aeiou" or word[:2] in ["yt", "xr"]:
            result.append(word + "ay")
            continue

        for idx in range(1, len(word)):
            if word[idx] in 'aeiou' + 'y':
                # break the consonant cluster with a vowel or 'y
                # Rule 4
                if word[idx-1] == 'q' and word[idx] == 'u':
                    # rule 3 for qu
                    idx += 1
                # rule 2
                result.append(word[idx:] + word[:idx] + 'ay')
                break

    return " ".join(result)

File: solutions/typescript/pig-latin/1/pig-latin.ts
export function translate(input: string): string {
  let result = ''
  let words = input.split(' ')
  for (let word of words) {
    let firstVowelIndex = word.search(/[aeiou]/)
    let firstLetter = word[0]
    if (firstVowelIndex === 0) {
      result += word + 'ay'
    } else if (firstLetter === 'q' && word[1] === 'u') {
      result += word.slice(2) + firstLetter + word[1] + 'ay'
    } else if (word[1] === 'q' && word[2] === 'u') {
      result += word.slice(3) + firstLetter + word[1] + word[2] + 'ay'
    } else {
      result += word.slice(firstVowelIndex) + word.slice(0, firstVowelIndex) + 'ay'
    }
    result += ' '
  }
  return result.trim()
}
